### DISCUSSION TITLE
[Bug] Switching to new OS breaks Scoop update

### DISCUSSION THREAD

**Discussion by quantuumsnot** [ScoopInstaller/Scoop#5386] [Help and Questions]
## Bug Report
Switching to new OS

#### Current Behavior
`scoop update` outputs this:

> fatal: detected dubious ownership in repository at 'D:/scoop/buckets/main'
> 'D:/scoop/buckets/main' is owned by:
>         'S-1-5-21-3453366618-1808197818-4029211567-1001'
> but the current user is:
>         'S-1-5-21-2567552669-4285415277-912380047-1001'
> To add an exception for this directory, call:
> 
>         git config --global --add safe.directory D:/scoop/buckets/main

And similar for each bucket/app
#### Expected Behavior
Normally user doesn't expect this to manually fix the package manager

### System details
**Windows version:** [10]

**OS architecture:** 64bit

**PowerShell version:** [output of `"$($PSVersionTable.PSVersion)"`]
> Major  5 Minor  1 Build  19041  Revision 2364

#### Scoop Configuration
If you say where is ~/.config/scoop/config.json in Win 10 I'll post it

**Comment by rashil2000** [comment]
What do you mean by "Switching to new OS"?

**Reply by quantuumsnot** [reply]
Well, it was and still works ok on Win 8.1 x64, switched to cleanly installed Win 10 to test the OS and scoop noped out (installed and used fully from a different non-OS drive)

**Reply by rashil2000** [reply]
Try readding the buckets from the new user:
```
scoop bucket rm main
scoop bucket add main
```

**Reply by quantuumsnot** [reply]
> scoop bucket add main
ERROR  is not a valid Git URL!
ERROR Please see https://git-scm.com/docs/git-clone#_git_urls for valid ones.
ERROR  is not a valid Git URL!
ERROR Please see https://git-scm.com/docs/git-clone#_git_urls for valid ones.
ERROR  is not a valid Git URL!
ERROR Please see https://git-scm.com/docs/git-clone#_git_urls for valid ones.
Checking repo... OK
The main bucket was added successfully.

Nope, same errors

**Comment by r15ch13** [comment]
Did you check the file permissions? If you changed to a new installed Windows it might not have write permissions.
`Rightclick > Properties > Security > Advanced > Owner: Change`

![image](https://user-images.githubusercontent.com/432127/219505464-6eb3dd0b-edfc-4a24-b0cf-31784bda63ad.png)

**Reply by quantuumsnot** [reply]
For the record scoop has full control permissions. Even reinstalled it again under current user with admin rights to try to fix the problem

**Comment by r15ch13** [comment]
Scoop config is located in `%USERPROFILE%\.config\scoop` or  `$env:USERPROFILE\.config\scoop`

What is the output of:
```pwsh
gci D:\scoop\buckets | % { $_.Name; pushd $_.FullName; git remote -v; popd }
```

Should look like this:
```pwsh
PS C:\Users\r15ch13> gci ~\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
extras
origin  https://github.com/ScoopInstaller/Extras (fetch)
origin  https://github.com/ScoopInstaller/Extras (push)
main
origin  https://github.com/ScoopInstaller/Main (fetch)
origin  https://github.com/ScoopInstaller/Main (push)
<snip>
```

**Reply by quantuumsnot** [reply]
`gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }`
Same error

**Reply by r15ch13** [reply]
If you run the suggested command `git config --global --add safe.directory D:/scoop/buckets/main` then it should work again.

I guess we have to address this new behavior of git.

**Reply by quantuumsnot** [reply]
Yeah, for all buckets in use :\

**Reply by r15ch13** [reply]
Adding the whole bucket folder should work too
`git config --global --add safe.directory D:/scoop/buckets`

**Reply by riotrah** [reply]
That last one didn't work for me

**Comment by quantuumsnot** [comment]
Btw this is the scoop's config.json

```
{
    "root_path":  "D:\\scoop",
    "global_path":  "D:\\scoop\\apps",
    "last_update":  "2023-02-19T18:38:04.0340847+02:00",
    "scoop_branch":  "master",
    "scoop_repo":  "https://github.com/ScoopInstaller/Scoop"
}

```

**Reply by r15ch13** [reply]
`global_path` both should point to a root directory for scoop.
e.g. `D:\scoop\global`, to get `D:\scoop\global\apps` and `D:\scoop\global\persist`

With your configuration you get `D:\scoop\apps\persist` and `D:\scoop\apps\apps`, was that your intention?

**Reply by quantuumsnot** [reply]
Nope, but apps are still loaded from `scoop\apps` and working ok. There's no subdir `apps` in `scoop\apps`

**Comment by quantuumsnot** [comment]
Forgot to post the output:

> What is the output of:
> 
> ```powershell
> gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
> ```


```
emulators
pushd : Cannot find path 'C:\Windows\system32\emulators' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\emulators:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
extras
pushd : Cannot find path 'C:\Windows\system32\extras' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\extras:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
games
pushd : Cannot find path 'C:\Windows\system32\games' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\games:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
main
pushd : Cannot find path 'C:\Windows\system32\main' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\main:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
qtools
pushd : Cannot find path 'C:\Windows\system32\qtools' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\qtools:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
versions
pushd : Cannot find path 'C:\Windows\system32\versions' because it does not exist.
At line:1 char:37
+ gci D:\scoop\buckets | % { $_.Name; pushd $_; git remote -v; popd }
+                                     ~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (C:\Windows\system32\versions:String) [Push-Location], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.PushLocationCommand

fatal: not a git repository (or any of the parent directories): .git
```

How should look `Environment Variables` for scoop and scooped apps under Win 10?

**Reply by r15ch13** [reply]
Change `pushd $_` to `pushd $_.FullName` (because of PowerShell 5)

You should have `D:\scoop\shims` in your `PATH` environment variable. Every other path can be restored with `scoop reset <app>`.

**Comment by quantuumsnot** [comment]
Ok, but just for the record I must say two things:

1. Copying or writing uppercase F with Shift + F doesn't work. For real can't add `.FullName` in PowerShell. Even copying the whole correct command trims `F` letter :D The user must restart PowerShell to be allowed ... WTF
2. What's the current expected behavior when resetting scoop apps?

Anyway, with your version:
```
emulators
fatal: not a git repository (or any of the parent directories): .git
extras
fatal: not a git repository (or any of the parent directories): .git
games
fatal: not a git repository (or any of the parent directories): .git
main
fatal: not a git repository (or any of the parent directories): .git
qtools
fatal: not a git repository (or any of the parent directories): .git
versions
fatal: not a git repository (or any of the parent directories): .git
```

**Reply by r15ch13** [reply]
1. That's strange 😕
2. `scoop reset` recreates the shim files, shortcuts, and env variables.

Just delete the buckets and add them again.
Does `git clone https://github.com/ScoopInstaller/Main` work at all?

**Reply by quantuumsnot** [reply]
Deleted and added the buckets which fixed the errors, but two of them were user created (in this case by me) and needed their repos to be added again. Is there an option that scoop can export and import it's own config and currently used buckets?

**Comment by riotrah** [comment]
Just wanted to add the following:

I have so many buckets I was reticent to rm & re-add them all. The issue in my case was that I installed scoop and all the buckets as an admin, but then changed my default terminal profile to a non-admin pwsh, causing the ownership issues in my case.

**Comment by 29039** [comment]
Duplicate of https://github.com/ScoopInstaller/Scoop/issues/4879#issuecomment-1793041043 where I have analysed the cause and offered a workaround script
